src/main/host/thread_ptrace.c

```diff
@@ -1079,41 +1079,63 @@ bool threadptrace_isRunning(Thread* base) {
     return false;
 }
 
-void threadptrace_terminate(Thread* base) {
-    ThreadPtrace* thread = _threadToThreadPtrace(base);
-
-    /* make sure we cleanup circular refs */
-    if (thread->sys) {
-        syscallhandler_unref(thread->sys);
-        thread->sys = NULL;
-    }
-
-    if (!threadptrace_isRunning(base)) {
+static void _threadptrace_terminateHelper(ThreadPtrace* thread) {
+    // Nothing to do if we already exited
+    if (!threadptrace_isRunning(&thread->base)) {
         return;
     }
 
-    // need to kill() and not ptrace() since the process may not be stopped
-    if (kill(thread->base.nativeTid, SIGKILL) < 0) {
-        warning("kill %d: %s", thread->base.nativePid, g_strerror(errno));
+    // First let's make sure that the thread is detached
+    if (!thread->needAttachment) {
+        // Still attached, detach before killing
+        info("Detaching attached native thread %d from pgroup %d now", thread->base.nativeTid,
+             thread->base.nativePid);
+        if (ptrace(PTRACE_DETACH, thread->base.nativeTid, 0, SIGSTOP) < 0) {
+            error("ptrace(DETACH) error %d while detaching from thread %d: %s", errno,
+                  thread->base.nativeTid, g_strerror(errno));
+        }
+        thread->needAttachment = true;
     }
 
-    // Wait for the exit ptrace-stop. Because the shadow process is the natural
-    // parent of the child (even if spawned from a task/thread other than this
-    // one), this also reaps the child.
-    int wstatus;
-    if (_waitpid_spin(thread->base.nativePid, &wstatus, 0) < 0) {
-        error("waitpid: %s", g_strerror(errno));
-        return;
+    // Already detached, kill the thread after making sure that it still exists
+    if (kill(thread->base.nativeTid, 0) == 0) {
```

Comment:
I don't think we need this check; we can just do the "real" kill and on error check for ESRCH, which would indicate it's already dead